libs/execution_context/ebpf_link.c

```diff
@@ -307,51 +358,54 @@ ebpf_link_detach_program(_Inout_ ebpf_link_t* link)
 {
     EBPF_LOG_ENTRY();
     ebpf_lock_state_t state;
-    ebpf_program_t* program = NULL;
-    bool link_is_detaching = false;
+    bool lock_held = false;
 
-    ebpf_object_acquire_reference((ebpf_core_object_t*)link);
+    ebpf_object_acquire_reference(&link->object);
 
-    state = ebpf_lock_lock(&link->attach_lock);
+    state = ebpf_lock_lock(&link->lock);
+    lock_held = true;
 
-    if (link->state == EBPF_LINK_STATE_ATTACHED || link->state == EBPF_LINK_STATE_ATTACHING) {
-        // This thread is responsible for detaching the link from the program.
-        _ebpf_link_set_state(link, EBPF_LINK_STATE_DETACHING);
-        link_is_detaching = true;
-        program = link->program;
-        ebpf_assert(program != NULL);
+    if (link->state != EBPF_LINK_STATE_ATTACHED) {
+        goto Done;
     }
 
-    if (link->state == EBPF_LINK_STATE_INITIAL || link->state == EBPF_LINK_STATE_DETACHED) {
-        ebpf_assert(link->program == NULL);
-    }
+    _ebpf_link_set_state(link, EBPF_LINK_STATE_DETACHING);
 
-    ebpf_lock_unlock(&link->attach_lock, state);
+    ebpf_lock_unlock(&link->lock, state);
+    lock_held = false;
 
-    if (!link_is_detaching) {
-        goto Done;
+    NTSTATUS status = NmrDeregisterClient(link->nmr_client_handle);
+    if (status == STATUS_PENDING) {
+        NmrWaitForClientDeregisterComplete(link->nmr_client_handle);
```

Comment:
Why block in this function?

Reply:
To ensure that when detach returns that the program is actually detached.

Reply:
We shouldn't care about ensuring the state when this function returns, only when the ioctl is completed.

Reply:
OK. That is a much larger change though. https://github.com/microsoft/ebpf-for-windows/issues/2456